mlir/lib/IR/AffineExpr.cpp

```diff
@@ -237,6 +239,144 @@ unsigned AffineDimExpr::getPosition() const {
+  if (exprKind == AffineExprKind::CeilDiv)
+    return 2;
+  return 0;
+}
+
+/// Returns true if the expression is divisible by the given symbol.
+static bool isDivisibleBySymbol(AffineExpr expr, unsigned symbolPos,
+                                int divideType) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+  case AffineExprKind::DimId:
+    return false;
+  case AffineExprKind::SymbolId:
+    return (expr.cast<AffineSymbolExpr>().getPosition() == symbolPos);
+  // Checking divisibility by given symbol for both operands.
+  case AffineExprKind::Add: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, divideType) &&
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, divideType);
+  }
+  // Operands of modulo if are floordiv are complicated to be simplify.
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, 0) &&
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, 0);
+  }
+  // Checking if any one of the operand divisible by given symbol.
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, divideType) ||
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, divideType);
+  }
+  // Floordiv and ceildiv are not divisible by symbol because of already
+  // existing division.
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::CeilDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    int type = getDivideType(expr.getKind());
+    if (divideType != type)
+      return false;
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, type);
+  }
+  }
+  llvm_unreachable("Unknown AffineExpr");
+}
+
+/// Divides a given expression by a given symbol. It considers all divisibility
+/// conditions are checked.
+static AffineExpr symbolicDivide(AffineExpr expr, unsigned symbolPos,
+                                 int divideType) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+  case AffineExprKind::DimId:
+    return nullptr;
+  case AffineExprKind::SymbolId:
+    return getAffineConstantExpr(1, expr.getContext());
+  // Dividing both operands by the given symbol.
+  case AffineExprKind::Add: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return getAffineBinaryOpExpr(
+        expr.getKind(),
+        symbolicDivide(binaryExpr.getLHS(), symbolPos, divideType),
+        symbolicDivide(binaryExpr.getRHS(), symbolPos, divideType));
+  }
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return getAffineBinaryOpExpr(
+        expr.getKind(), symbolicDivide(binaryExpr.getLHS(), symbolPos, 0),
+        symbolicDivide(binaryExpr.getRHS(), symbolPos, 0));
```

Comment:
The '0' argument like this make the code unreadable. You need to either use an enum or just send the `AffineExprKind`.